dynamiqs/qarray.py

```diff
@@ -0,0 +1,558 @@
+        """Convert the quantum state to a density matrix.
+
+        Returns:
+            The density matrix representation of the quantum state.
+        """
+
+    def toop(self) -> QArray:
+        """Convert the quantum state to an operator.
+
+        Returns:
+            The operator representation of the quantum state.
+        """
+        return self.todm()
+
+    def proj(self) -> QArray:
+        """Projector of the quantum state.
+
+        Returns:
+            The projector of the quantum state.
+        """
+        return self.todm()
+
+    # conversion methods
+    @abstractmethod
+    def to_numpy(self) -> np.ndarray:
+        """Convert the quantum state to a NumPy array.
+
+        Returns:
+            The NumPy array representation of the quantum state.
+        """
+
+    @abstractmethod
+    def to_qutip(self) -> qt.QObj:
+        """Convert the quantum state to a QuTiP object.
+
+        Returns:
+            A QuTiP object representation of the quantum state.
+        """
+
+    @abstractmethod
+    def to_jax(self) -> Array:
+        """Convert the quantum state to a JAX array.
+
+        Returns:
+            The JAX array representation of the quantum state.
+        """
+
+    def __repr__(self) -> str:
+        return (
+            f'{type(self).__name__}(shape={self.shape}, '
+            f'dims={self.dims}, dtype={self.dtype})'
+        )
+
+    def __str__(self) -> str:
+        return self.__repr__()
+
+    def __neg__(self) -> QArray:
+        """Negate the quantum state."""
+        return -1 * self
+
+    def __mul__(
+        self, other: ScalarLike | ArrayLike @ abstractmethod
+    ) -> QArray:  # warning if used with array
+        """Element-wise multiplication with a scalar or an array."""
+
+    def __rmul__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with array
+        """Element-wise multiplication with a scalar or an array on the right."""
+
+    def __add__(
+        self, other: ScalarLike | ArrayLike @ abstractmethod
+    ) -> QArray:  # warning if used with scalar
+        """Element-wise addition with a scalar or an array."""
+
+    def __radd__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with scalar
+        """Element-wise addition with a scalar or an array on the right."""
+        return self + other
+
+    def __sub__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with scalar
+        """Element-wise subtraction with a scalar or an array."""
+
+    def __rsub__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with scalar
+        """Element-wise subtraction with a scalar or an array on the right."""
+        return -self + other
+
+    @abstractmethod
+    def __matmul__(self, other: ArrayLike) -> QArray:
+        """Matrix multiplication with another quantum state or JAX array."""
+
+    def __rmatmul__(self, other: ArrayLike) -> QArray:
```

Comment:
Abstract too, matmul is not commutative so needs to be implemented twice by the child class.